### PR TITLE
#217 add basic architecture for benchmarking using jmh

### PR DESCRIPTION
**Pull request by julien-truffaut**

### CODE REVIEW

**Comment by julien-truffaut** [comment]
```
sbt
> project bench
> run -i 10 -wi 3 -f3

[info] Benchmark                       (size)  Mode  Cnt     Score     Error  Units
[info] QueryParamBench.setQueryParams      10  avgt   30     0.812 ±   0.023  us/op
[info] QueryParamBench.setQueryParams     100  avgt   30     9.060 ±   0.250  us/op
[info] QueryParamBench.setQueryParams    1000  avgt   30    97.872 ±   1.876  us/op
[info] QueryParamBench.withQueryParam      10  avgt   30     1.655 ±   0.070  us/op
[info] QueryParamBench.withQueryParam     100  avgt   30    75.717 ±   1.437  us/op
[info] QueryParamBench.withQueryParam    1000  avgt   30  8126.523 ± 126.294  us/op
```

File: bench/src/main/scala/org/http4s/bench/QueryParamBench.scala [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import java.util.concurrent.TimeUnit
+
+import org.openjdk.jmh.annotations._
+
+@BenchmarkMode(Array(Mode.AverageTime))
+@OutputTimeUnit(TimeUnit.MICROSECONDS)
+@State(Scope.Benchmark)
+class QueryParamBench {
+
+  @Benchmark def lensGet0(in: Nested0Input) =
```
Comment (bryce-anderson) [inline bench/src/main/scala/org/http4s/bench/QueryParamBench.scala]: Did a commit get missed? This appears to be incomplete and is the source of the Travis failure.

Comment (julien-truffaut) [inline bench/src/main/scala/org/http4s/bench/QueryParamBench.scala]: yes probably, I will check tonight

Comment (julien-truffaut) [inline bench/src/main/scala/org/http4s/bench/QueryParamBench.scala]: really weird, it seems that my last commit got lost. I will add it back

**Comment by julien-truffaut** [comment]
should be ok now

**Comment by rossabaker** [comment]
Excellent.  This should help us add some rigor when we start talking performance.

**Pull request merged**